### PR TITLE
Create backups files without having to copy inner tarballs

### PR DESCRIPTION
**Pull request by bdraco**
~~needs https://github.com/pvizeli/securetar/pull/33~~
~~needs https://github.com/home-assistant/supervisor/pull/4888~~
core version https://github.com/home-assistant/core/pull/110267

## Proposed change

This is a followup to #4843 where I noticed a large chunk of the time on I/O bound systems is copying the data into the tarfile and than making another tarfile of the original tarfiles. This will allow us to effectively cut the disk writes in half when creating backups and increase storage media longevity.  Additionally this means the user no longer needs twice as much disk space as the backup size since there is only one copy of the backup being written to disk now.

To avoid the double copy, we now write each tarfile into the fileobj of the outer tar file.

This reduced my backup time on my fast system from 24s to 10s, on I/O bound systems, the reduction is multiple minutes.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to the supervisor)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:
- Link to cli pull request:

## Checklist

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Ruff (`ruff format supervisor tests`)
- [ ] Tests have been added to verify that the new code works.

If API endpoints of add-on configuration are added/changed:

- [ ] Documentation added/updated for [developers.home-assistant.io][docs-repository]

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[docs-repository]: https://github.com/home-assistant/developers.home-assistant

### CODE REVIEW

File: tests/backups/test_backup.py [hunk 1]
```diff
@@ -16,7 +16,7 @@ async def test_new_backup_stays_in_folder(coresys: CoreSys, tmp_path: Path):
 
     async with backup:
         assert len(listdir(tmp_path)) == 1
-        assert not backup.tarfile.exists()
+        assert backup.tarfile.exists()
```
Comment (bdraco) [inline tests/backups/test_backup.py:19]: We now create the backup file right away so we can stream into it.

File: supervisor/backups/backup.py [hunk 2]
```diff
@@ -356,34 +386,33 @@ async def __aexit__(self, exception_type, exception_value, traceback):
...
         tar_name = f"{addon.slug}.tar{'.gz' if self.compressed else ''}"
-        addon_file = SecureTarFile(
-            Path(self._tmp.name, tar_name),
-            "w",
-            key=self._key,
+
+        addon_file = self._outer_secure_tarfile.create_inner_tar(
+            f"./{tar_name}",
             gzip=self.compressed,
-            bufsize=BUF_SIZE,
```
Comment (bdraco) [inline supervisor/backups/backup.py:384]: bufsize is now set in the outer tar and the inner tar uses the bufsize of the outer tar

File: supervisor/backups/backup.py [hunk 3]
```diff
@@ -677,12 +707,12 @@ async def store_homeassistant(self, exclude_database: bool = False):
             ATTR_EXCLUDE_DATABASE: exclude_database,
         }
 
+        tar_name = f"homeassistant.tar{'.gz' if self.compressed else ''}"
         # Backup Home Assistant Core config directory
-        tar_name = Path(
-            self._tmp.name, f"homeassistant.tar{'.gz' if self.compressed else ''}"
-        )
-        homeassistant_file = SecureTarFile(
-            tar_name, "w", key=self._key, gzip=self.compressed, bufsize=BUF_SIZE
```
Comment (bdraco) [inline supervisor/backups/backup.py:685]: bufsize is now set in the outer tar and the inner tar uses the bufsize of the outer tar

**Comment by bdraco** [comment]
verified broken addon backups still generate a successful backup (the lib has tests for this but I wanted to do it manually to be sure)

Feb 12 18:53:46 homeassistant hassio_supervisor[490]: 24-02-12 12:53:46 ERROR (MainThread) [supervisor.addons.addon] Can't write tarfile <securetar._InnerSecureTarFile object at 0xffff8abb19a0>: [Errno 2] No such file or directory: '/data/addon_configs/a0d7b954_nodered'

**Review by agners — APPROVED** [review]

**Comment by bdraco** [comment]
Thanks. Deployed wider and everything still looks good.

**Pull request merged**